UnwrapBlog.md

```diff
@@ -0,0 +1,92 @@
+
+We knew that providing the implementation such a function itself was not going to the true hard part.  After all, we already had the code to peel back one layer.  So peeling back all the layers was more a matter of just having the function call itself recursively when encountering things like arrays (or other objects that themselves contained `Input`s).  The hard part came in when were trying figure out how would we even express this in TypeScript?.  After all, what's the type signature when you effectively to say:
+
+```ts
+function unwrap(input: Input<...MaybeHasInputsOrPromisesOrOutputsArbitrarilyDeeper...>): SomeThingWithAllTheInputsPromisesAndOutputsErased;
+```
+
+This was a big headscratcher for a while.  But after exploring some of the powerful new features that TypeScript has been bringing about in recent releases, we finally found a great way to manage it.  In the end, it ends up looking something like this:
+
+```ts
+type primitive = string | number | boolean | undefined | null;
+
+type Unwrap<T> =
+    T extends Promise<infer U> ? UnwrapSimple<U> :
+    T extends Output<infer U> ? UnwrapSimple<U> :
+    UnwrapSimple<T>;
+
+type UnwrapSimple<T> =
+    T extends primitive ? T :
+    T extends Array<infer U> ? UnwrappedArray<U> :
+    T extends object ? UnwrappedObject<T> :
+    never;
+
+interface UnwrappedArray<T> extends Array<Unwrap<T>> {}
+
+type UnwrappedObject<T> = {
+    [P in keyof T]: Unwrap<T[P]>;
+};
+
+function unwrap<T>(val: T): Unwrap<T>;
+```
+
+It's a bit of a doozy containing a bunch of type definitions and the final function declaration that uses them.  It makes use of many advanced TypeScript typing features.  Including 'Union Types' (the 'primitive' type), 'Mapped Types' (which 'UnwrappedObject' is an example of), 'Conditional Types' (the types with `?` and `:` in them), 'Inferred Types' (the use of `infer U` in several places).  You can read more about these in the [Advanced Types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) documentation.
+
+Put together, the `Unwrap<T>` is effectively itself a recursively defined structural pattern matching type.  You can see it reference itself once you walk into `UnwrappedArray<T>` and `UnwrappedObject<T>` types. As it recurses, it matches the type it has and then 'peels away' the wrappings that have been caused by `Promise<T>` and `Output<T>`.  As it hits Arrays and Objects, it dives deeper, unwrapping along the way.  This is all manageable in TypeScript because it does this sort of matching and unwrapping in a lazy fashion, allowing us to express what is effectively an infinitely large pattern-matching type, without it having any trouble understanding it or applying it!
+
+It amazes us how well this works and how amazing TypeScript is here.  As you can see here, TS completely understands what's going on and will give you all the great type-checking support and tooling that we know and love:
```

Comment:
Nit, we should try to be consistent -- either it's TypeScript or TS... I prefer TypeScript.